src/analyses/abortUnless.ml

```diff
@@ -0,0 +1,70 @@
+(** An analysis checking whether a function only returns if its only argument has a non-zero value. *)
+
+open Prelude.Ana
+open Analyses
+
+module Spec =
+struct
+  include Analyses.DefaultSpec
+
+  let name () = "abortUnless"
+  module D = BoolDomain.MustBool
+  module C = Lattice.Unit
+
+  let context _ _ = ()
+
+  (* transfer functions *)
+  let assign ctx (lval:lval) (rval:exp) : D.t =
+    false
+
+  let branch ctx (exp:exp) (tv:bool) : D.t =
+    ctx.local
+
+  let body ctx (f:fundec) : D.t =
+    ctx.local
+
+  let return ctx (exp:exp option) (f:fundec) : D.t =
+    if ctx.local then
+      match f.sformals with
+      | [arg] when isIntegralType arg.vtype ->
+        (match ctx.ask (EvalInt (Lval (Var arg, NoOffset))) with
+         | v when Queries.ID.is_bot v -> false
+         | v ->
+           match Queries.ID.to_bool v with
+           | Some b -> b
+           | None -> false)
+      | _ ->
+        (* should not happen, ctx.local should always be false in this case *)
+        false
+    else
+      false
+
+  let enter ctx (lval: lval option) (f:fundec) (args:exp list) : (D.t * D.t) list =
+    let candidate = match f.sformals with
+      | [arg] when isIntegralType arg.vtype -> true
```

Comment:
This allows calls with similar signature to be called within `assume_abort_if_not`, including (mutually) recursively. It's not obvious why something bad cannot come from that.

Reply:
Nice catch! It's not the recursion though that was problematic, as we go to `false` in `combine` either way but the fact that the assert happens after evaluation of the transfer function for call, so if variables in `arg` are assigned to, asserting might unsoundly yield bot (as in the case you constructed).